File: 003. 无重复字符的最长子串.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        left = 0
        max_str = 0
        window = set()  # window表示以s[right]结尾的最长无重复字符的串
        for right in range(len(s)):
            while s[right] in window:
                window.remove(s[left])
                left += 1
            window.add(s[right])
            max_str = max(max_str, len(window))
        return max_str


s = Solution()
print(s.lengthOfLongestSubstring("abcabcbb"))

File: 003. 无重复字符的最长子串.go
package main

func lengthOfLongestSubstring(s string) int {
	// 右指针从左往右扫,扫到的字符都放到滑动窗口里,
	// 如果扫到的字符已经在窗口里了，就删除滑动窗口里的这个字符并不断右移左指针直到无重复

	if len(s) == 0 {
		return 0
	}

	res := 0
	l := 0
	r := 0
	window := make(map[byte]struct{}, 0)

	for r < len(s) {
		if _, exist := window[s[r]]; exist {
			delete(window, s[l])
			l++
		} else {
			window[s[r]] = struct{}{}
			r++
			res = max(res, r-l)
		}
	}

	return res
}
